keysign/bluetoothoffer.py

```diff
@@ -0,0 +1,143 @@
+import logging
+from bluetooth import BluetoothSocket, RFCOMM
+import dbus
+import select
+import socket
+if __name__ == "__main__":
+    import gi
+    gi.require_version('Gtk', '3.0')
+    from twisted.internet import gtk3reactor
+    gtk3reactor.install()
+    from twisted.internet import reactor
+from twisted.internet import threads
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+if __name__ == "__main__" and __package__ is None:
+    logging.getLogger().error("You seem to be trying to execute " +
+                              "this script directly which is discouraged. " +
+                              "Try python -m instead.")
+    import os
+    parent_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    os.sys.path.insert(0, parent_dir)
+    os.sys.path.insert(0, os.path.join(parent_dir, 'monkeysign'))
+    import keysign
+    __package__ = str('keysign')
+
+from .gpgmh import get_public_key_data, get_usable_keys
+from .util import get_local_bt_address, mac_generate
+
+log = logging.getLogger(__name__)
+
+
+class BluetoothOffer:
+    def __init__(self, key, port=3, size=1024):
+        self.key = key
+        self.port = port
+        self.size = size
+        self.server_socket = None
+        self.message_def = None
+        self.stopped = False
+        self.code = None
+
+    @inlineCallbacks
+    def start(self):
+        self.stopped = False
+        message = "Back"
+        success = False
+        try:
+            while not self.stopped and not success:
+                # server_socket.accept() is not stoppable. So with select we can call accept()
+                # only when we are sure that there is already a waiting connection
+                ready_to_read, ready_to_write, in_error = yield threads.deferToThread(
+                    select.select, [self.server_socket], [], [], 0.5)
+                if ready_to_read:
+                    # We are sure that a connection is available, so we can call
+                    # accept() without deferring it to a thread
+                    client_socket, address = self.server_socket.accept()
+                    key_data = get_public_key_data(self.key.fingerprint)
+                    kd_decoded = key_data.decode('utf-8')
+                    yield threads.deferToThread(client_socket.sendall, kd_decoded)
```

Comment:
> Hm. True. Good point. But do we currently care about when we have finished sending the key? I don't see any consumer of the start function using the return value.

This one. Even if we don't check the result we need to know that the pending operation is ended in order to start a new sending.
https://github.com/gnome-keysign/gnome-keysign/pull/18/files/46d99329e4ec49d2b6aa602f9437e8cc6cb0c669#diff-554ea0006765f03a4424c0c27a5550bcR105

> We could probably replace that select easily with hooking it up to glib if we wanted to, though

Yeap that's an option

> I'm talking about the client_socket

Ops, I was looking at server_socket 😄